File: Algorithms/DLX/Ruby/dlx.rb
#!/usr/bin/env ruby
# DLX (Dancing Links) Sudoku Solver - Ruby Implementation
# Algorithm X with Dancing Links for Exact Cover problem
#
# This implements Knuth's Algorithm X using circular doubly-linked lists
# for efficient column covering/uncovering operations.

# DLX Node - circular doubly-linked list node
class DlxNode
  attr_accessor :up, :down, :left, :right, :column, :row_id

  def initialize
    @up = self
    @down = self
    @left = self
    @right = self
    @column = nil
    @row_id = -1
  end
end

# DLX Column - special node that also tracks column size
class DlxColumn < DlxNode
  attr_accessor :size, :name

  def initialize(name = "")
    super()
    @size = 0
    @name = name
    @column = self  # Column points to itself
  end
end

# Global iteration counter (analogous to brute-force count)
$dlx_iterations = 0

# Sudoku puzzle and solution
$puzzle = Array.new(9) { Array.new(9, 0) }
$solution_grid = Array.new(9) { Array.new(9, 0) }

# DLX matrix structures
$root = nil
$columns = []
$row_info = []  # Array of {row:, col:, num:} hashes

# Cover a column in the DLX matrix
# Remove column from header list and remove all rows in the column's list
def cover_column(c)
  # Remove column header from the header list
  c.right.left = c.left
  c.left.right = c.right

  # For each row in this column
  row_node = c.down
  while row_node != c
    # For each node in this row (excluding the column itself)
    right_node = row_node.right
    while right_node != row_node
      # Remove this node from its column
      right_node.down.up = right_node.up
      right_node.up.down = right_node.down
      right_node.column.size -= 1
      right_node = right_node.right
    end
    row_node = row_node.down
  end
end

# Uncover a column (exact reverse of cover)
def uncover_column(c)
  # For each row in this column (in reverse order)
  row_node = c.up
  while row_node != c
    # For each node in this row (in reverse order)
    left_node = row_node.left
    while left_node != row_node
      # Restore this node to its column
      left_node.column.size += 1
      left_node.down.up = left_node
      left_node.up.down = left_node
      left_node = left_node.left
    end
    row_node = row_node.up
  end

  # Restore column header to the header list
  c.right.left = c
  c.left.right = c
end

# Choose column with minimum size (Knuth's S heuristic)
def choose_column(root)
  best = nil
  min_size = Float::INFINITY

  col_node = root.right
  while col_node != root
    if col_node.size < min_size
      min_size = col_node.size
      best = col_node
    end
    col_node = col_node.right
  end

  best
end

# DLX Search - Algorithm X with Dancing Links
# Returns true if solution found, false otherwise
def dlx_search(root, k, solution)
  $dlx_iterations += 1  # Count every search call

  # If matrix is empty, we found a solution
  return true if root.right == root

  # Choose column with minimum size
  col = choose_column(root)

  # If column has no rows, no solution possible
  return false if col.size == 0

  # Cover this column
  cover_column(col)

  # Try each row in this column
  row_node = col.down
  while row_node != col
    # Add row to partial solution
    solution[k] = row_node.row_id

    # Cover all other columns in this row
    right_node = row_node.right
    while right_node != row_node
      cover_column(right_node.column)
      right_node = right_node.right
    end

    # Recurse
    if dlx_search(root, k + 1, solution)
      return true  # Solution found
    end

    # Backtrack: uncover all columns in this row
    left_node = row_node.left
    while left_node != row_node
      uncover_column(left_node.column)
      left_node = left_node.left
    end

    row_node = row_node.down
  end

  # Uncover column
  uncover_column(col)

  false  # No solution found
end

# Calculate constraint column indices
# For a Sudoku cell (r,c) with value n (1-9):
# - Position constraint: (r,c) must be filled -> column index: r*9 + c
# - Row constraint: row r must have number n -> column index: 81 + r*9 + (n-1)
# - Column constraint: column c must have number n -> column index: 162 + c*9 + (n-1)
# - Box constraint: box b must have number n -> column index: 243 + b*9 + (n-1)

def get_position_col(r, c)
  r * 9 + c
end

def get_row_col(r, n)
  81 + r * 9 + (n - 1)
end

def get_col_col(c, n)
  162 + c * 9 + (n - 1)
end

def get_box_col(r, c, n)
  box = (r / 3) * 3 + (c / 3)
  243 + box * 9 + (n - 1)
end

# Add a node to the DLX matrix
def add_node(col, row_id)
  node = DlxNode.new
  node.column = col
  node.row_id = row_id

  # Insert at end of column's circular list
  node.down = col
  node.up = col.up
  col.up.down = node
  col.up = node
  col.size += 1

  node
end

# Build a DLX row for Sudoku cell (r,c) with value n
def build_dlx_row(r, c, n, row_id)
  # Store row metadata
  $row_info[row_id] = {row: r, col: c, num: n}

  # Create nodes for the 4 constraints
  n1 = add_node($columns[get_position_col(r, c)], row_id)
  n2 = add_node($columns[get_row_col(r, n)], row_id)
  n3 = add_node($columns[get_col_col(c, n)], row_id)
  n4 = add_node($columns[get_box_col(r, c, n)], row_id)

  # Link nodes horizontally in circular list
  n1.right = n2
  n2.right = n3
  n3.right = n4
  n4.right = n1

  n1.left = n4
  n2.left = n1
  n3.left = n2
  n4.left = n3

  n1
end

# Initialize DLX matrix structure
def init_dlx_matrix
  # Allocate root column
  $root = DlxColumn.new("root")

  # Allocate 324 column headers
  $columns = []
  324.times do |i|
    col = DlxColumn.new("C#{i}")

    # Link into header list
    col.left = $root.left
    col.right = $root
    $root.left.right = col
    $root.left = col

    $columns << col
  end

  $row_info = []
end

# Build the complete DLX matrix from the puzzle
def build_dlx_matrix_from_puzzle
  row_id = 0
  row_starts = []

  9.times do |r|
    9.times do |c|
      if $puzzle[r][c] != 0
        # Cell has a clue - create only one row for that value
        row_starts[row_id] = build_dlx_row(r, c, $puzzle[r][c], row_id)
        row_id += 1
      else
        # Cell is empty - create rows for all possible values
        (1..9).each do |n|
          row_starts[row_id] = build_dlx_row(r, c, n, row_id)
          row_id += 1
        end
      end
    end
  end

  row_starts
end

# Cover given clues (pre-selected rows)
def cover_clues(row_starts)
  9.times do |r|
    9.times do |c|
      if $puzzle[r][c] != 0
        n = $puzzle[r][c]

        # Find the row for this clue
        row_starts.each_with_index do |start_node, row_id|
          next unless start_node
          info = $row_info[row_id]
          if info[:row] == r && info[:col] == c && info[:num] == n
            # Cover all columns in this row
            node = start_node
            curr = node
            loop do
              cover_column(curr.column)
              curr = curr.right
              break if curr == node
            end
            break
          end
        end
      end
    end
  end
end

# Extract solution from DLX and populate solution_grid
def extract_solution(solution)
  # Initialize solution grid with original puzzle
  $solution_grid = $puzzle.map(&:dup)

  # Each solution entry is a row_id
  solution.each do |row_id|
    next if row_id.nil? || row_id < 0
    info = $row_info[row_id]
    $solution_grid[info[:row]][info[:col]] = info[:num] if info
  end
end

# Print puzzle
def print_puzzle(grid)
  puts "\nPuzzle:"
  grid.each do |row|
    puts row.map(&:to_s).join(" ") + " "
  end
end

# Read matrix file
def read_matrix_file(filename)
  # Normalize path for output
  display_path = filename
  if filename.start_with?("/app/Matrices/")
    display_path = "../" + filename[5..-1]
  end
  puts display_path

  line_count = 0
  File.foreach(filename) do |line|
    line = line.strip
    next if line.empty? || line.start_with?('#')

    values = line.split.map(&:to_i)
    if values.length == 9 && line_count < 9
      $puzzle[line_count] = values
      puts values.map(&:to_s).join(" ") + " "
      line_count += 1
    end
  end
end

# Main program
if ARGV.empty?
  puts "Usage: ruby dlx.rb <matrix_file>"
  exit 1
end

start_time = Time.now

ARGV.each do |filename|
  next unless filename.end_with?(".matrix")

  # Reset puzzle
  $puzzle = Array.new(9) { Array.new(9, 0) }

  read_matrix_file(filename)
  print_puzzle($puzzle)

  # Initialize DLX matrix
  init_dlx_matrix

  # Build matrix from puzzle
  row_starts = build_dlx_matrix_from_puzzle

  # Cover pre-filled clues
  cover_clues(row_starts)

  # Solve using DLX
  $dlx_iterations = 0
  solution = Array.new(81)
  result = dlx_search($root, 0, solution)

  if result
    extract_solution(solution)
    print_puzzle($solution_grid)
    puts "\nSolved in Iterations=#{$dlx_iterations}\n"
  else
    puts "\nNo solution found"
  end
end

elapsed = Time.now - start_time
puts "Seconds to process %.3f" % elapsed

File: Algorithms/DLX/PHP/dlx.php
#!/usr/bin/env php
<?php
declare(strict_types=1);

// ============================================================================
// GLOBAL ITERATION COUNTER
// ============================================================================
$dlx_iterations = 0;

// ============================================================================
// DLX CORE FUNCTIONS
// ============================================================================

function dlx_cover_column($col) {
    // Remove column header from the header list
    $col->right->left = $col->left;
    $col->left->right = $col->right;

    // For each row in this column
    $row = $col->down;
    while ($row !== $col) {
        // For each node in this row (excluding the column itself)
        $right = $row->right;
        while ($right !== $row) {
            // Remove this node from its column
            $right->down->up = $right->up;
            $right->up->down = $right->down;
            $right->column->size--;
            $right = $right->right;
        }
        $row = $row->down;
    }
}

function dlx_uncover_column($col) {
    // For each row in this column (in reverse order)
    $row = $col->up;
    while ($row !== $col) {
        // For each node in this row (in reverse order)
        $left = $row->left;
        while ($left !== $row) {
            // Restore this node to its column
            $left->column->size++;
            $left->down->up = $left;
            $left->up->down = $left;
            $left = $left->left;
        }
        $row = $row->up;
    }

    // Restore column header to the header list
    $col->right->left = $col;
    $col->left->right = $col;
}

function choose_column($root) {
    $best = null;
    $min_size = PHP_INT_MAX;

    $col = $root->right;
    while ($col !== $root) {
        if ($col->size < $min_size) {
            $min_size = $col->size;
            $best = $col;
        }
        $col = $col->right;
    }

    return $best;
}

function dlx_search($root, $k, &$solution) {
    global $dlx_iterations;
    $dlx_iterations++;

    // If matrix is empty, we found a solution
    if ($root->right === $root) {
        return true;
    }

    // Choose column with minimum size
    $col = choose_column($root);

    // If column has no rows, no solution possible
    if ($col === null || $col->size == 0) {
        return false;
    }

    // Cover this column
    dlx_cover_column($col);

    // Try each row in this column
    $row = $col->down;
    while ($row !== $col) {
        // Add row to partial solution
        $solution[$k] = $row->row_id;

        // Cover all other columns in this row
        $right = $row->right;
        while ($right !== $row) {
            dlx_cover_column($right->column);
            $right = $right->right;
        }

        // Recurse
        if (dlx_search($root, $k + 1, $solution)) {
            return true;
        }

        // Backtrack: uncover all columns in this row
        $left = $row->left;
        while ($left !== $row) {
            dlx_uncover_column($left->column);
            $left = $left->left;
        }

        $row = $row->down;
    }

    // Uncover column
    dlx_uncover_column($col);

    return false;
}

// ============================================================================
// SUDOKU-SPECIFIC FUNCTIONS
// ============================================================================

function get_position_col($r, $c) {
    return $r * 9 + $c;
}

function get_row_col($r, $n) {
    return 81 + $r * 9 + ($n - 1);
}

function get_col_col($c, $n) {
    return 162 + $c * 9 + ($n - 1);
}

function get_box_col($r, $c, $n) {
    $box = intdiv($r, 3) * 3 + intdiv($c, 3);
    return 243 + $box * 9 + ($n - 1);
}

function init_dlx_matrix() {
    // Create root column
    $root = new stdClass();
    $root->left = $root;
    $root->right = $root;
    $root->up = $root;
    $root->down = $root;
    $root->size = 0;
    $root->name = 'root';

    // Create 324 column headers
    $columns = [];
    for ($i = 0; $i < 324; $i++) {
        $col = new stdClass();
        $col->up = $col;
        $col->down = $col;
        $col->size = 0;
        $col->name = "C$i";

        // Link into header list
        $col->left = $root->left;
        $col->right = $root;
        $root->left->right = $col;
        $root->left = $col;

        $columns[$i] = $col;
    }

    return ['root' => $root, 'columns' => $columns];
}

function add_node($dlx, $col_idx, $row_id) {
    $col = $dlx['columns'][$col_idx];

    $node = new stdClass();
    $node->column = $col;
    $node->row_id = $row_id;

    // Insert at end of column's circular list
    $node->down = $col;
    $node->up = $col->up;
    $col->up->down = $node;
    $col->up = $node;
    $col->size++;

    return $node;
}

function build_dlx_row($dlx, $r, $c, $n, $row_id, &$row_info) {
    // Store row metadata
    $row_info[$row_id] = ['row' => $r, 'col' => $c, 'num' => $n];

    // Create nodes for the 4 constraints
    $n1 = add_node($dlx, get_position_col($r, $c), $row_id);
    $n2 = add_node($dlx, get_row_col($r, $n), $row_id);
    $n3 = add_node($dlx, get_col_col($c, $n), $row_id);
    $n4 = add_node($dlx, get_box_col($r, $c, $n), $row_id);

    // Link nodes horizontally in circular list
    $n1->right = $n2;
    $n1->left = $n4;

    $n2->right = $n3;
    $n2->left = $n1;

    $n3->right = $n4;
    $n3->left = $n2;

    $n4->right = $n1;
    $n4->left = $n3;

    return $n1;
}

function build_exact_cover_matrix($puzzle) {
    $dlx = init_dlx_matrix();
    $row_info = [];
    $row_starts = [];
    $row_id = 0;

    for ($r = 0; $r < 9; $r++) {
        for ($c = 0; $c < 9; $c++) {
            if ($puzzle[$r][$c] != 0) {
                // Cell has a clue - create only one row for that value
                $row_starts[$row_id] = build_dlx_row($dlx, $r, $c, $puzzle[$r][$c], $row_id, $row_info);
                $row_id++;
            } else {
                // Cell is empty - create rows for all possible values
                for ($n = 1; $n <= 9; $n++) {
                    $row_starts[$row_id] = build_dlx_row($dlx, $r, $c, $n, $row_id, $row_info);
                    $row_id++;
                }
            }
        }
    }

    return ['dlx' => $dlx, 'row_info' => $row_info, 'row_starts' => $row_starts, 'puzzle' => $puzzle];
}

function cover_clues($data) {
    $puzzle = $data['puzzle'];
    $row_info = $data['row_info'];
    $row_starts = $data['row_starts'];

    for ($r = 0; $r < 9; $r++) {
        for ($c = 0; $c < 9; $c++) {
            if ($puzzle[$r][$c] != 0) {
                $n = $puzzle[$r][$c];

                // Find the row for this clue
                foreach ($row_starts as $row_id => $node) {
                    if (isset($row_info[$row_id]) &&
                        $row_info[$row_id]['row'] == $r &&
                        $row_info[$row_id]['col'] == $c &&
                        $row_info[$row_id]['num'] == $n) {

                        // Cover all columns in this row
                        $curr = $node;
                        do {
                            dlx_cover_column($curr->column);
                            $curr = $curr->right;
                        } while ($curr !== $node);
                        break;
                    }
                }
            }
        }
    }
}

function extract_solution($data, $solution) {
    $puzzle = $data['puzzle'];
    $row_info = $data['row_info'];
    $solution_grid = $puzzle;

    foreach ($solution as $row_id) {
        if (isset($row_info[$row_id])) {
            $info = $row_info[$row_id];
            $solution_grid[$info['row']][$info['col']] = $info['num'];
        }
    }

    return $solution_grid;
}

function print_puzzle($grid) {
    echo "\nPuzzle:\n";
    for ($r = 0; $r < 9; $r++) {
        for ($c = 0; $c < 9; $c++) {
            echo $grid[$r][$c] . " ";
        }
        echo "\n";
    }
}

function parse_matrix($filename) {
    if (!file_exists($filename)) {
        fwrite(STDERR, "Error opening file '$filename'\n");
        return null;
    }

    // Normalize path for output
    $display_path = $filename;
    if (strpos($filename, '/app/Matrices/') === 0) {
        $display_path = '../' . substr($filename, 5);
    }
    echo "$display_path\n";

    $lines = file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    $puzzle = [];
    $line_count = 0;

    foreach ($lines as $line) {
        $line = trim($line);
        if (empty($line) || $line[0] === '#') {
            continue;
        }

        $parts = preg_split('/\s+/', $line);
        if (count($parts) >= 9) {
            $row = [];
            for ($i = 0; $i < 9; $i++) {
                $row[] = (int)$parts[$i];
                echo $parts[$i] . " ";
            }
            echo "\n";
            $puzzle[] = $row;
            $line_count++;

            if ($line_count == 9) {
                break;
            }
        }
    }

    if ($line_count != 9) {
        fwrite(STDERR, "Error: file does not contain 9 lines\n");
        return null;
    }

    return $puzzle;
}

// ============================================================================
// MAIN
// ============================================================================

if ($argc < 2) {
    fwrite(STDERR, "Usage: php dlx.php <file1> <file2> ...\n");
    exit(1);
}

$start_time = microtime(true);

for ($i = 1; $i < $argc; $i++) {
    $filename = $argv[$i];

    if (substr($filename, -7) !== '.matrix') {
        continue;
    }

    $puzzle = parse_matrix($filename);
    if ($puzzle === null) {
        continue;
    }

    print_puzzle($puzzle);

    // Build DLX matrix
    $data = build_exact_cover_matrix($puzzle);

    // Cover pre-filled clues
    cover_clues($data);

    // Solve using DLX
    $dlx_iterations = 0;
    $solution = [];
    $result = dlx_search($data['dlx']['root'], 0, $solution);

    if ($result) {
        $solution_grid = extract_solution($data, $solution);
        print_puzzle($solution_grid);
        echo "\nSolved in Iterations=$dlx_iterations\n\n";
    } else {
        echo "\nNo solution found\n\n";
    }
}

$end_time = microtime(true);
$elapsed = $end_time - $start_time;
printf("Seconds to process %.3f\n", $elapsed);
